aggregator/shard.go

```diff
@@ -49,16 +72,48 @@ func newAggregatorShard(shard uint32, opts Options) *aggregatorShard {
 	// its own time lock to ensure for an aggregation window, all metrics
 	// owned by the shard are aggregated before they are flushed.
 	opts = opts.SetTimeLock(&sync.RWMutex{})
+	scope := opts.InstrumentOptions().MetricsScope().SubScope("shard").Tagged(
+		map[string]string{"shard": strconv.Itoa(int(shard))},
+	)
 	s := &aggregatorShard{
-		shard:     shard,
-		metricMap: newMetricMap(shard, opts),
+		shard: shard,
+		nowFn: opts.ClockOptions().NowFn(),
+		bufferDurationBeforeShardCutover: opts.BufferDurationBeforeShardCutover(),
+		bufferDurationAfterShardCutoff:   opts.BufferDurationAfterShardCutoff(),
+		metricMap:                        newMetricMap(shard, opts),
+		metrics:                          newAggregatorShardMetrics(scope),
 	}
 	s.addMetricWithPoliciesListFn = s.metricMap.AddMetricWithPoliciesList
 	return s
 }
 
 func (s *aggregatorShard) ID() uint32 { return s.shard }
 
+func (s *aggregatorShard) SetWriteableRange(cutoverNanos, cutoffNanos int64) {
+	earliestNanos := int64(0)
+	if cutoverNanos >= int64(s.bufferDurationBeforeShardCutover) {
+		earliestNanos = cutoverNanos - int64(s.bufferDurationBeforeShardCutover)
+	}
+
+	latestNanos := int64(math.MaxInt64)
+	if cutoffNanos <= math.MaxInt64-int64(s.bufferDurationAfterShardCutoff) {
+		latestNanos = cutoffNanos + int64(s.bufferDurationAfterShardCutoff)
+	}
+
+	s.Lock()
+	s.earliestWritableNanos = earliestNanos
+	s.latestWriteableNanos = latestNanos
+	s.Unlock()
+}
+
+func (s *aggregatorShard) IsCutoff() bool {
+	nowNanos := s.nowFn().UnixNano()
```

Comment:
Do we want to get `nowNanos` while holding the lock to ensure calls to `IsCutoff` are synchronized with respect to calls to `AddMetricWithPoliciesList`?

Reply:
Don't think that's necessary, also this is called from the ticking goroutine so even if we hold a write lock here, once the call returns the goroutine no longer holds the lock so it's no different from the current implementation.